### DISCUSSION TITLE
Is it possible to configure a selection filter by annotation

### DISCUSSION THREAD

**Discussion by mikepictor** [TimefoldAI/timefold-solver#859] [General]
I don't use a config XML, all my configuration is via annotations.

I want to add a selection filter, but I can't find any annotation to flag it's inclusion. Is the XML config file the only way to include selection filters? The documentation seems to only describe XML configuration.

Thank you

**Comment by triceo** [comment]
Not that I know of. You can configure a pinning filter on `@PlanningEntity`, but not a generic selection filter.

Do note though that XML is not the only way of configuring the solver. In fact, the XML is just a serialized version of `SolverConfig` - so if you start from that class, you can build your config in code without any XML.

**Reply by mikepictor** [reply]
Oh...ok, let me look at that. I can maybe initialize a selection filter in the bootup process. 

Thanks.

**Reply by mikepictor** [reply]
is it possible to get the existing solverConfig, and just amend it? If I create my own, I have to set all the values currently being managed by annotation (EG solutionClass), which just feels like I am reinventing the wheel. If I can get a handle on the existing, I can just inject my filter, though even at that...I am not sure where in the config the filter is meant to go

Boy it would be nice if I could just say `@PlanningSolution(filter = MyFilter::class)`...so long as I'm wishing...

If I could crack this nut, I think I could get some real performance gains because I have a few specific "no go" scenarios that I'd love to head off at the pass

**Reply by triceo** [reply]
Since you speak of injection, I'm assuming you're using either Quarkus or Spring Boot. In that case, yes, you can have `SolverConfig` injected, just like you can inject a `SolverManager` etc. That said, to get as deep as to selection filters (which are typically on move selectors), you may still need to do quite some heavy lifting.

Selection filters are not very nice to use, and they are also not good performance-wise - in order to apply selection filters, selection first needs to happen; you'll be selecting something only to immediately throw it away. The ideal way of dealing with selection filters is not using selection filters; either have these "no go" scenarios enforced directly by the model (not allowing impossible states to be modelled), or use value providers on entities, which allows you to eliminate impossible scenarios even before selection. (But if you're using list variables, value providers on entites are not supported.)

In the end, you may be forced to use selection filters, and that will in turn force you to figure out how to configure them. It is by no means ideal, which is why selection filters are not public API; they are a workaround for a feature gap in the solver. A feature gap which we will be addressing, but not in the short term.

**Reply by mikepictor** [reply]
Huh..yes..SpringBoot in my case

In this case I am not sure I can do it with the value providers, as the hard limit is variable (it's the arrival of a physical truck, scheduling a task to come before that time is just pointless....but different tasks, different trucks, different "no go" timeframes)

As it is, I just created hard constraints, and it works, but I know I am incurring processing time (including for other constraints) as it samples various times before that truck. I was hoping to use the filter to say "any task that starts before the truck arrival time...start over"

Basically I want valid time "windows"...but different tasks will have different valid windows.

Maybe I can't (easily) get there from here. My solving works, I just know that the model is testing various scenarios where those times are early enough to be invalid, and then getting a bad score which is why I don't see them in the end. I was hoping to really take a bite out of useless computation time.

**Reply by triceo** [reply]
Yes, I see your point. In fact, value ranges on entities are there to serve this exact purpose:

- If the truck "knows" the earliest availability, then that is the start of the value range.
- If the truck's earliest availability is based on some shadow variable (such as arrival of some other truck), then that shadow variable is the start of the value range.
- The end of the value range is harder to figure out; but you can make an educated guess. For example, if you only expect trucks to be out for 8 hours at most, make the end be `start + 8 + safety margin`.

By applying these rules, you get the range from `[0, end of planning horizon + safety margin]` to something significantly smaller, therefore significantly reducing the "useless" computation time. But as I said before, ranges on entities are not supported with list variables and since you're mentioning trucks, you're likely using those.

**Reply by mikepictor** [reply]
Yes, I came up with an idea. I had the value range provider in the solution, but I noticed we could have it on the entity, and so each entity can have a value range unique to it's truck time. I am going to try that. Takes some re-architecting, but it could remove many hours from the "eligible" times of the Temporal value range. That will feel like a better solution if I can get my ducks in a row.

**Reply by mikepictor** [reply]
Pinned entities still need a value range provider?

I got an error saying that the range provider on my pinned entity was null...but it's pinned. It should need a value range. I mean...I guess I can set one, but that was very unexpected.

**Reply by triceo** [reply]
I guess the code that checks for that doesn't also check if the entity is pinned or not. I can see how that could be a bug. But since pinning is a runtime thing, I can also see the logic behind enforcing value ranges on _all_ entities; if the entity becomes unpinned, you don't want the solver to suddenly fail as it encounters an unpinned entity without a value range.

**Reply by mikepictor** [reply]
yeah...in this case it was an version of the entity that is always pinned. I have multiple sub-classes, 2 of which are perma pinned, the rest can be planned.

It's fine, I gave it a value range of 12:00 to 12:05 and called it a day.

It's working, thank you for your help. I now have all tasks using their own value range based on the truck arrival time. That should save a lot of useless time ranges.
